toolkit/featureforms/src/main/java/com/arcgismaps/toolkit/featureforms/components/text/FormTextFieldState.kt

```diff
@@ -102,52 +114,135 @@ internal class FormTextFieldState(
+    
+    /**
+     * The domain of the element's field.
+     */
+    val domain: Domain? = properties.domain
+    
+    /**
+     * The FieldType of the element's field.
+     */
+    val fieldType: FieldType = properties.fieldType
+    
     private var _errorMessage: String = ""
 
     // build helper text
-    private val helperText = if (minLength > 0 && maxLength > 0) {
-        if (minLength == maxLength) {
-            context.getString(R.string.enter_n_chars, minLength)
+    private val helperText =
+        if (fieldType.isNumeric) {
+            if (domain != null && domain is RangeDomain) {
+                val min = domain.minValue
+                val max = domain.maxValue
+                // to format the range of either integer or floating point
+                // values without a lot of logic, they are formatted as strings.
+                if (min is Number && max is Number) {
+                    context.getString(R.string.numeric_range_helper_text, min.format(), max.format())
+                } else if (min is Number) {
+                     context.getString(R.string.less_than_min_value, min.format())
+                } else if (max is Number) {
+                    context.getString(R.string.exceeds_max_value, max.format())
+                } else {
+                    // not likely to happen.
+                    ""
+                }
+            } else {
+                ""
+            }
         } else {
-            context.getString(R.string.enter_min_to_max_chars, minLength, maxLength)
+            if (minLength > 0 && maxLength > 0) {
+                if (minLength == maxLength) {
+                    context.getString(R.string.enter_n_chars, minLength)
+                } else {
+                    context.getString(R.string.enter_min_to_max_chars, minLength, maxLength)
+                }
+            } else if (maxLength > 0) {
+                context.getString(R.string.maximum_n_chars, maxLength)
+            } else {
+                context.getString(R.string.maximum_n_chars, 254)
+            }
         }
-    } else if (maxLength > 0) {
-        context.getString(R.string.maximum_n_chars, maxLength)
-    } else {
-        context.getString(R.string.maximum_n_chars, 254)
-        // TODO: when consuming the core API throw here and remove the line above.
-        //throw IllegalStateException("invalid form data or attribute: text field must have a nonzero max length")
-    }
 
     init {
         scope.launch {
             value.drop(1).collect { newValue ->
                 if (isEditable.value) {
-                    validate(newValue)
+                    validate(newValue, true)
                 }
             }
         }
         scope.launch {
             isFocused.collect {
                 if (it) {
-                    validate(value.value)
+                    validate(value.value, true)
+                } else {
+                    validate(value.value, !isRequired.value)
+                }
```

Comment:
```suggestion
                    validate(value.value, !isRequired.value)
                }
```
Is there a reason for this change? This would preemptively show all validation errors even for fields that cannot be edited.

Reply:
fields that cannot be edited are disabled and cannot be focused. Their focus state will never change. If the field is focused, it should not show the "Required" error string according to the design spec.

Reply:
I have added a change to drop the first value collected from the isFocused StateFlow, so validation doesn't occur preemptively.